mrmustard/lab_dev/samplers.py

```diff
@@ -0,0 +1,229 @@
+            reduced_state = (state >> meas_op).normalize()
+            samples = self.sample(reduced_state, counts)
+            for sample in samples:
+                ret.append(np.append([unique_sample], sample))
+        return np.array(ret)
+
+    def sample_prob_dist(
+        self, state: State, n_samples: int = 1000, seed: int | None = None
+    ) -> np.ndarray:
+        r"""
+        Samples a state by computing the probability distribution.
+
+        Args:
+            state: The state to sample.
+            n_samples: The number of samples to generate.
+            seed: An optional seed for random sampling.
+        """
+        rng = np.random.default_rng(seed) if seed else settings.rng
+        return rng.choice(
+            a=list(product(self.meas_outcomes, repeat=len(state.modes))),
+            p=self.probabilities(state),
+            size=n_samples,
+        )
+
+    def _get_povm(self, meas_outcome: Any, mode: int) -> CircuitComponent:
+        r"""
+        Returns the POVM associated with a given outcome on a given mode.
+
+        Args:
+            meas_outcome: The measurement outcome.
+            mode: The mode.
+
+        Returns:
+            The POVM circuit component.
+        """
+        if isinstance(self.povms, CircuitComponent):
+            kwargs = self.povms.parameter_set.to_dict()
+            kwargs[self._outcome_arg] = meas_outcome
+            return self.povms.__class__(modes=[mode], **kwargs)
+        else:
+            return self.povms[self.meas_outcomes.index(meas_outcome)].on([mode])
+
+    def _validate_probs(self, probs: Sequence[float], atol: float) -> Sequence[float]:
+        r"""
+        Validates that the given probability distribution sums to ``1`` within some
+        tolerance and returns a renormalized probability distribution to account for
+        small numerical errors.
+
+        Args:
+            probs: The probability distribution to validate.
+            atol: The absolute tolerance to validate with.
+        """
+        atol = atol or settings.ATOL
+        prob_sum = sum(probs)
+        if not math.allclose(prob_sum, 1, atol):
+            raise ValueError(f"Probabilities sum to {prob_sum} and not 1.0.")
+        return math.real(probs / prob_sum)
+
+
+class PNRSampler(Sampler):
+    r"""
+    A sampler for photon-number resolving (PNR) detectors.
+
+    Args:
+        cutoff: The photon number cutoff.
+    """
+
+    def __init__(self, cutoff: int) -> None:
+        super().__init__(list(range(cutoff)), Number([0], 0))
+        self._cutoff = cutoff
+        self._outcome_arg = "n"
+
+    def probabilities(self, state, atol=1e-4):
+        return self._validate_probs(state.fock_distribution(self._cutoff), atol)
+
+
+class HomodyneSampler(Sampler):
+    r"""
+    A sampler for homodyne measurements.
+
+    Args:
+        phi: The quadrature angle where ``0`` corresponds to ``x`` and ``\pi/2`` to ``p``.
```

Comment:
Perhaps we should leave this at 0 by default (homodyne always on x quadrature) and let people put an Rgate before to reach any angle that way.

Reply:
yeah I take it back, it's probably simpler this way.